File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/22-GenerateParentheses/22-GenerateParentheses.java
// Last updated: 8/4/2026, 7:26:00 PM
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String>list = new ArrayList<>();
        generate(n , 0 , 0 , "" , list);
        return list;
    }
    public static void generate(int n , int op , int cl , String ans , List<String>ll){
        if(op == n && cl == n){
            ll.add(ans);
            return;
        }
        if(op < n){
            generate(n , op + 1 , cl , ans + "(" , ll);
        }
        if(op > cl){
            generate(n , op , cl + 1 , ans + ")" , ll);
        }

    }
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/22-GenerateParentheses/22-GenerateParentheses.cpp
// Last updated: 8/4/2026, 7:25:54 PM
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string>ans;
        solve(0, 0 ,"", n ,  ans);
        return ans;
    }
    void solve(int op ,int cl,  string s ,int n ,vector<string>& ans){
        
        if(op == n && cl == n){
            ans.push_back(s);
            return;
        }

        if(op < n){
            solve(op+1 , cl , s+"(" , n , ans);
        }
        if(cl < op){
            solve(op , cl+1 , s + ")" , n , ans);
        }
    }
};